File: api/R/gdh.r
# Depends: httr readr
# (openssl and tibble are installed with the above packages)
# These packages are part of tidyverse (tidyverse.org), so
# if you have that installed, you don't need to install anything.

# Example invocation:
# > source("gdh.r")
# > key <- "API KEY HERE"
# > c1 <- get_cases(key, country = "New Zealand")
#
# Use get_cached_cases() to cache the cases locally. This is useful for
# rerunning the script which will then use the cached version.

library(httr)
library(tibble)
library(readr)
library(openssl)

GDH_URL_DEFAULT <- "https://data.covid-19.global.health"
GDH_URL <- if((url <- Sys.getenv("GDH_URL")) != "")
    url else GDH_URL_DEFAULT
downloadAsync <- "/api/cases/downloadAsync"

# sync with data-serving/data-service/src/util/search.ts
FILTERS <- c(
    "country"
)

# query string function
stringify_filters <- function(...) {
    kwargs <- list(...)
    fields <- names(kwargs)
    for(f in fields) {
        if( !(f %in% FILTERS)) {
            stop(paste("Invalid field entered:", f))
        }
    }
    for(field in fields) {
        if(grepl(" ", kwargs[[field]])) {
            kwargs[[field]] = paste0("'", kwargs[[field]], "'")
        }
    }
    for(date_field in c("dateconfirmedafter", "dateconfirmedbefore")) {
        if(date_field %in% fields && inherits(kwargs[[date_field]], 'Date')) {
            kwargs[[date_field]] = paste(kwargs[[date_field]])  # convert to string
        }
    }
    querystring <- ""
    for(field in fields) {
        querystring <- paste0(querystring, " ", field, ":", kwargs[[field]])
    }
    return(trimws(querystring))
}

get_cases <- function(apikey, server = GDH_URL, ...) {
    data <- sprintf('{"format": "csv", "query": "%s"}', trimws(stringify_filters(...)))
    res <- POST(paste0(server, downloadAsync), body = data,
                add_headers("Content-Type" = "application/json",
                            "X-API-Key" = apikey))
    if(res$status_code != 200) {
        stop(content(res))
    }
    t <- content(res)
    if(is_tibble(t)) {
        return(t)
    } else {
        # country-export returns a gzip compressed file
        con <- gzcon(url(t$signedUrl))
        return(read_csv(con))
    }
}

cases_cachefile <- function(server = GDH_URL, folder="cache", ...) {
    return(paste0(folder, "/",
                  sha256(paste0(stringify_filters(...), "|", server)), ".csv"))
}

# Returns a cached copy of cases if it exists, otherwise saves to cache
get_cached_cases <- function(apikey, server = GDH_URL, refresh = FALSE, folder = "cache", ...) {
    if(!file.exists(folder)) {
        dir.create(folder)
    }
    cachename <- cases_cachefile(server, folder, ...)
    if(!refresh && file.exists(cachename)) {
        return(read_csv(cachename))
    } else {
        df <- get_cases(apikey, server, ...)
    }
    write_csv(df, cachename)
    return(df)
}

File: api/python/gdh.py
# Depends: requests pandas
# You may already have these installed, otherwise install using pip:
#     pip install requests pandas

# Example invocations:
# >>> from gdh import GlobalDotHealth
# >>> key = "API KEY HERE"
# >>> c19 = GlobalDotHealth(key, 'covid-19')
# >>> c1 = c19.get_cases(country="New Zealand")
#
# Use get_cached_cases() to cache the cases locally. This is useful for
# rerunning the script which will then use the cached version.

import os
import io
import json
import hashlib
import datetime

import requests
import pandas as pd

downloadAsync = "/api/cases/downloadAsync"

# sync with data-serving/data-service/src/util/search.ts
FILTERS = {
    "country"
}


def stringify_filters(**kwargs):
    if not kwargs:
        raise ValueError("At least one filter has to be specified")
    if unknown_filters := set(kwargs) - FILTERS:
        raise ValueError(f"Unknown filters: {unknown_filters}")
    kwargs = {
        field: (f"'{value}'" if (isinstance(value, str) and " " in value) else value)
        for field, value in kwargs.items()
    }
    for date_field in ["dateconfirmedafter", "dateconfirmedbefore"]:
        # convert to yyyy-mm-dd if in datetime.date format
        if date_field in kwargs and isinstance(
            kwargs[date_field], datetime.date
        ):
            kwargs[date_field] = kwargs[date_field].isoformat()
    return " ".join(f"{field}:{value}" for field, value in kwargs.items())


def cases_cachefile(server, folder="cache", **kwargs):
    to_hash = f"{stringify_filters(**kwargs)}|{server}"
    sha256 = hashlib.sha256(to_hash.encode("utf-8")).hexdigest()
    return f"{folder}/{sha256}.csv"


class GlobalDotHealth:
    def __init__(self, apikey, disease='covid-19'):
        # Let someone override our URL explicitly
        if server := os.getenv('GDH_URL'):
            self.server = server
        else:
            # Otherwise, build a production URL from the disease name
            self.server = f'https://data.{disease}.global.health'
        self.apikey = apikey

    def get_cases(self, **kwargs):
        res = requests.post(
            f"{self.server}{downloadAsync}",
            data=json.dumps({"format": "csv", "query": stringify_filters(**kwargs)}),
            headers={"Content-Type": "application/json", "X-API-Key": self.apikey},
        )
        if res.status_code != 200:
            raise ConnectionError(res.text)
        if "signedUrl" in res.text:
            # country-export returns a gzip compressed file
            signedUrl = json.loads(res.text)["signedUrl"]
            return pd.read_csv(signedUrl, compression="gzip")
        else:
            with io.StringIO(res.text) as buf:
                return pd.read_csv(buf)

    # Returns a cached copy of cases if it exists, otherwise saves to cache
    def get_cached_cases(refresh=False, folder="cache", **kwargs):
        if not os.path.exists(folder):
            os.mkdir(folder)
        cachename = cases_cachefile(self.server, folder, **kwargs)
        if not refresh and os.path.exists(cachename):
            return pd.read_csv(cachename)
        else:
            df = self.get_cases(**kwargs)
        df.to_csv(cachename, index=False)
        return df
